### PR TITLE
lokiexporter: Implementation ( PR 2/3 )

### PR DESCRIPTION
**Pull request by gramidt**
Description:
The implementation of the Loki exporter.

Link to tracking Issue:
#1894

Testing:
Unit Tests

Documentation:
Godoc and README

Notes:
#1952 was the original PR, but we were unable to re-open it after it was closed for being stale.

### CODE REVIEW

**Comment by gramidt** [comment]
Happy Tuesday, @kbrockhoff! 

Hope your week is off to a great start! Do you have a timeframe when you would be able to review this? No rush, just need to ensure that I block out time to respond promptly during the review process. 

Thank you for all of your help in the Otel community!

File: exporter/lokiexporter/example/otel-collector-config.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  health_check:
+  pprof:
+  zpages:
+
+service:
+  extensions: [pprof, zpages, health_check]
+  pipelines:
+    logs:
+      receivers: [ fluentforward ]
+      processors: [ batch ]
```
Comment (kbrockhoff) [inline exporter/lokiexporter/example/otel-collector-config.yml:28]: Adding memorylimiter to processor list would likely help noobs

File: exporter/lokiexporter/exporter.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	resp, err := l.client.Do(req)
+	if err != nil {
+		numDroppedLogs += ld.LogRecordCount()
+		return numDroppedLogs, err
+	}
+
+	_, _ = io.Copy(ioutil.Discard, resp.Body)
+	_ = resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
```
Comment (kbrockhoff) [inline exporter/lokiexporter/exporter.go:96]: Retries and other resiliency handling is expected to be implemented in an exporter. exporterhelper package in opentelemetry-collector project has functionality for assisting in this.

**Review by kbrockhoff — CHANGES_REQUESTED** [review]

File: exporter/lokiexporter/example/otel-collector-config.yml (hunk 1 above)
Comment (gramidt) [reply exporter/lokiexporter/example/otel-collector-config.yml:28]: Excellent suggestion!

File: exporter/lokiexporter/exporter.go (hunk 2 above)
Comment (gramidt) [reply exporter/lokiexporter/exporter.go:96]: Great catch, @kbrockhoff! I forgot to add those back in.

Comment (gramidt) [reply exporter/lokiexporter/exporter.go:94]: Updated with the appropriate consumererrors so that the export helper would retry when appropriate.

**Comment by cyriltovena** [comment]
Nice work @gramidt appreciate the effort 👏 

We're starting a new repo for easing client integration here https://github.com/grafana/loki-client-go , I'm not saying you should use this, but just wanted to make you aware :) it would simplify your go.mod and how much vendor you have to pull into this repository.

How does the log push request batching is working ? Sorry to ask I'm not familiar with open-telemetry log collector.

**Comment by gramidt** [comment]
@cyriltovena - Thank you so much! I'm so glad to see that a Loki client was split out. This will definitely help with the inevitable dependency issue. I will go ahead and use the protos from 'loki-client-go' now, then submit a future PR with updating the code to use the other pieces of the client where applicable.

The OpenTelemetry Collector has the concept of Receivers, Processors, and Exporters. Receivers are responsible for getting data from systems (push or pull), Processors are responsible for processing the data (Batching, Queueing, Sampling, etc), and Exporters are responsible for sending the data off to a another system to either continue processing or storing the data. Here is the list of the "core" processors: https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/README.md . There are also additional "contrib" processors in this repo.

**Comment by gramidt** [comment]
@cyriltovena - After further look, I will switch over to 'loki-client-go' for the protos after an official release of it is out. Looking forward to when it's available. :)

**Review by kbrockhoff — APPROVED** [review]

**Comment by gramidt** [comment]
Thank you for all of your help, @kbrockhoff!

**Comment by gramidt** [comment]
@bogdandrutu - When you're available, could you merge this?

**Pull request merged**